src/nvapi_d3d12.cpp

```diff
@@ -240,4 +240,146 @@ extern "C" {
 
         return Ok(n, alreadyLoggedOk);
     }
+
+    NvAPI_Status __cdecl NvAPI_D3D12_GetRaytracingCaps(ID3D12Device* pDevice, NVAPI_D3D12_RAYTRACING_CAPS_TYPE type, void* pData, size_t dataSize) {
+        constexpr auto n = __func__;
+        static bool alreadyLoggedOk = false;
+
+        if (pDevice == nullptr || pData == nullptr)
+            return InvalidPointer(n);
+
+        switch (type) {
+            case NVAPI_D3D12_RAYTRACING_CAPS_TYPE_THREAD_REORDERING:
+                if (dataSize != sizeof(NVAPI_D3D12_RAYTRACING_THREAD_REORDERING_CAPS))
+                    return InvalidArgument(n);
+
+                // let's hope that NvAPI_D3D12_IsNvShaderExtnOpCodeSupported returning false is enough to discourage games from attempting to use SER
+                *(NVAPI_D3D12_RAYTRACING_THREAD_REORDERING_CAPS*)pData = NVAPI_D3D12_RAYTRACING_THREAD_REORDERING_CAP_NONE;
+                break;
+
+            case NVAPI_D3D12_RAYTRACING_CAPS_TYPE_OPACITY_MICROMAP:
+                if (dataSize != sizeof(NVAPI_D3D12_RAYTRACING_OPACITY_MICROMAP_CAPS))
+                    return InvalidArgument(n);
+
+                *(NVAPI_D3D12_RAYTRACING_OPACITY_MICROMAP_CAPS*)pData = NVAPI_D3D12_RAYTRACING_OPACITY_MICROMAP_CAP_NONE;
+                break;
+
+            default:
+                return InvalidArgument(n);
+        }
+
+        return Ok(n, alreadyLoggedOk);
+    }
+
+    static NvAPI_Status ConvertBuildRaytracingAccelerationStructureInputs(const NVAPI_D3D12_BUILD_RAYTRACING_ACCELERATION_STRUCTURE_INPUTS_EX* nvDesc, std::vector<D3D12_RAYTRACING_GEOMETRY_DESC>& geometryDescs, D3D12_BUILD_RAYTRACING_ACCELERATION_STRUCTURE_INPUTS* d3dDesc) {
+        *d3dDesc = {
+            .Type = nvDesc->type,
+            // assume that OMM via VK_EXT_opacity_micromap is not supported, allow only standard flags to be passed
+            .Flags = static_cast<D3D12_RAYTRACING_ACCELERATION_STRUCTURE_BUILD_FLAGS>(nvDesc->flags & 0x3f),
+            .NumDescs = nvDesc->numDescs,
+            .DescsLayout = nvDesc->descsLayout,
+            .InstanceDescs = nvDesc->instanceDescs,
+        };
+
+        if (d3dDesc->Type == D3D12_RAYTRACING_ACCELERATION_STRUCTURE_TYPE_BOTTOM_LEVEL && d3dDesc->DescsLayout == D3D12_ELEMENTS_LAYOUT_ARRAY) {
+            if (nvDesc->geometryDescStrideInBytes != sizeof(NVAPI_D3D12_RAYTRACING_GEOMETRY_DESC_EX))
+                return NVAPI_INCOMPATIBLE_STRUCT_VERSION;
```

Comment:
I guess `NVAPI_INVALID_ARGUMENT` fits better. `NVAPI_INCOMPATIBLE_STRUCT_VERSION` should I guess be reserved for struct  version mismatches.
(This also removes one `case` from the two switch statements below)

Reply:
If there's only one type of error reported, I could also return a `bool` instead and replace the entire `switch` with a simpler `if`. Should I do that or would you like me to keep `NvAPI_Status` there?

Reply:
Good point, lets go for `bool`, we already do that in other places e.g.  `NvapiD3d11Device`